src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -532,6 +533,94 @@ uct_ib_mlx5_devx_open_device(struct ibv_device *ibv_device,
     return ctx;
 }
 
+#ifdef HAVE_DECL_MLX5DV_DEVX_SUBSCRIBE_DEVX_EVENT
+ucs_status_t uct_ib_mlx5_devx_subscribe_event(uct_ib_mlx5_md_t *md,
+                                              struct mlx5dv_devx_obj *obj,
+                                              unsigned event_num,
+                                              unsigned event_type,
+                                              unsigned event_data)
+{
+    uint64_t cookie;
+    uint16_t event;
+    int ret;
+
+    if (!md->super.dev.async_events) {
+        return UCS_OK;
+    }
+
+    event  = event_num;
+    cookie = event_type | (event_data << UCT_IB_MLX5_DEVX_EVENT_DATA_SHIFT);
+    ret    = mlx5dv_devx_subscribe_devx_event(md->event_channel, obj,
+                                              sizeof(event), &event, cookie);
+    if (ret) {
+        ucs_error("mlx5dv_devx_subscribe_devx_event() failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+static void uct_ib_mlx5_devx_event_handler(int fd, void *arg)
+{
+    struct mlx5dv_devx_async_event_hdr devx_event;
+    uct_ib_async_event_t event;
+    uct_ib_mlx5_md_t *md = arg;
+    int ret;
+
+    ret = mlx5dv_devx_get_event(md->event_channel, &devx_event, sizeof(devx_event));
+    if (ret < 0) {
+        ucs_warn("mlx5dv_devx_get_event() failed: %m");
+        return;
+    }
+
+    event.event_type = devx_event.cookie & UCT_IB_MLX5_DEVX_EVENT_TYPE_MASK;
+    switch (event.event_type) {
+    case IBV_EVENT_QP_LAST_WQE_REACHED:
+        event.qp_num = devx_event.cookie >> UCT_IB_MLX5_DEVX_EVENT_DATA_SHIFT;
+        break;
+    }
+
+    uct_ib_handle_async_event(&md->super.dev, &event);
+}
+
+static ucs_status_t uct_ib_mlx5_devx_init_events(uct_ib_mlx5_md_t *md)
+{
+    if (!md->super.dev.async_events) {
+        return UCS_OK;
+    }
+
+    md->event_channel = mlx5dv_devx_create_event_channel(
+            md->super.dev.ibv_context,
+            MLX5_IB_UAPI_DEVX_CR_EV_CH_FLAGS_OMIT_DATA);
+
+    if (md->event_channel == NULL) {
+            ucs_error("mlx5dv_devx_create_event_channel() failed: %m");
+            return UCS_ERR_IO_ERROR;
+    }
+
+    return ucs_async_set_event_handler(UCS_ASYNC_THREAD_LOCK_TYPE,
```

Comment:
1. set the fd to nonblocking
2. create completion channel per iface, not per MD, so it will be on the async context which belongs to the UCT worker